packages/zen-bull-vault/src/EmergencyWithdraw.sol

```diff
@@ -0,0 +1,253 @@
+     * @param _zenBullAmount ZenBull amount to redeem
+     * @param _maxEthForWPowerPerp max ETH to pay for flashswapped oSQTH amount
+     */
+    function emergencyWithdrawEthFromCrab(uint256 _zenBullAmount, uint256 _maxEthForWPowerPerp)
+        external
+    {
+        uint256 crabToRedeem = _zenBullAmount.wdiv(zenBullTotalSupplyForCrabWithdrawal).wmul(
+            IZenBullStrategy(zenBull).getCrabBalance()
+        );
+        (, uint256 wPowerPerpInCrab) = IZenBullStrategy(zenBull).getCrabVaultDetails();
+        uint256 wPowerPerpToRedeem =
+            crabToRedeem.wmul(wPowerPerpInCrab).wdiv(IERC20(crab).totalSupply());
+
+        zenBullTotalSupplyForCrabWithdrawal =
+            zenBullTotalSupplyForCrabWithdrawal.sub(_zenBullAmount);
+
+        IERC20(zenBull).transferFrom(msg.sender, address(this), _zenBullAmount);
+        _mint(msg.sender, _zenBullAmount);
+
+        _exactOutFlashSwap(
+            weth,
+            wPowerPerp,
+            3000,
+            wPowerPerpToRedeem,
+            _maxEthForWPowerPerp,
+            uint8(FLASH_SOURCE.EMERGENCY_WITHDRAW_ETH_FROM_CRAB),
+            abi.encodePacked(crabToRedeem, wPowerPerpToRedeem)
+        );
+
+        uint256 payout = IERC20(weth).balanceOf(address(this));
+        IWETH9(weth).withdraw(payout);
+        payable(msg.sender).sendValue(payout);
+
+        emit Withdraw(
+            msg.sender, _zenBullAmount, crabToRedeem, wPowerPerpToRedeem, payout, _zenBullAmount
+        );
+    }
+
+    /**
+     * @notice repay a portion of Euler debt and withdraw WETH collateral to this contract based on ZenBullEulerRecovery amount
+     * @dev this will revert of WETH to withdraw from Euler is greater than max, or limit price breach the tolerance %
+     * @dev if all ZenBullEulerRecovery are burnt, the ETH withdrawl will be activated
+     * @param _wethToWithdraw amount of WETH to withdraw from Euler
+     * @param _maxEthForUsdc max ETH to pay for 1 USDC
+     * @param _poolFee ETH/USDC Uni v3 pool fee
+     */
+    function emergencyRepayEulerDebt(
+        uint256 _wethToWithdraw,
+        uint256 _maxEthForUsdc,
+        uint24 _poolFee
+    ) external {
+        require(
+            _wethToWithdraw <= MAX_WETH_PER_DEBT_REPAY,
+            "WETH to withdraw is greater than max per repay"
+        );
+
+        uint256 ratio = _wethToWithdraw.wdiv(IEulerEToken(eToken).balanceOfUnderlying(zenBull));
+        uint256 usdcToRepay = ratio.wmul(IEulerDToken(dToken).balanceOf(zenBull));
+        uint256 ethUsdPrice = UniOracle._getTwap(ethUSDCPool, weth, usdc, 420, false);
+
+        require(
+            _maxEthForUsdc <= ethUsdPrice.wmul((ONE.add(LIMIT_PRICE_TOLERANCE))),
+            "Max ETH for USDC greater than max price tolerance"
+        );
+
+        _exactOutFlashSwap(
+            weth,
+            usdc,
+            _poolFee,
+            usdcToRepay,
+            usdcToRepay.wmul(_maxEthForUsdc.mul(1e12)),
```

Comment:
maybe make magic # more explicit

Reply:
this should be usdcToRepay.mul(1e12).wdiv(_maxEthForUsdc), right?  its an amount of ETH in max

Reply:
ie 1000e6 (1k$) assuming 1k$ eth price -> 1e6*1e12/1000e18 = 1eth max in

Reply:
But maxEthForUsd is how much ETH to pay for 1 USDC right? at least that's how I did it

Reply:
```UniOracle._getTwap(ethUSDCPool, weth, usdc, 420, false);``` returns USDC per 1 ETH scaled to 1e18

```        require(
            _maxEthForUsdc <= ethUsdPrice.wmul((ONE.add(LIMIT_PRICE_TOLERANCE))),
            "Max ETH for USDC greater than max price tolerance" ```
            
          
This also takes USDC per ETH prices.

Reply:
I updated it to take USDC price per 1 ETH